Lib/fnmatch.py

```diff
@@ -86,10 +97,29 @@ def translate(pat):
         i = i+1
         if c == '*':
             # compress consecutive `*` into one
-            if (not res) or res[-1] is not STAR:
+            h = i-1
+            while i < n and pat[i] == '*':
+                i = i+1
+            if sep:
+                star_count = i-h
+                is_segment = (h == 0 or pat[h-1] == sep) and (i == n or pat[i] == sep)
+                if star_count == 1:
+                    if is_segment:
+                        add(f'{DOT}+')
```

Comment:
Prior to `sep`, `*` translates to `.*` not `.+`:

```
>>> fnmatch.translate('*')
'(?s:.*)\\Z'
```

And that's the behavior I see in shell too:

```
@ ls *README.rst
README.rst
```

I'm thinking `glob('*')` should match zero characters.

Reply:
When it appears as its own segment, `*` consumes at least one character in both pathlib and `glob.glob()`:

```python
>>> import os, glob
>>> os.listdir()
['empty']
>>> os.listdir('empty/')
[]
>>> glob.glob('empty/*')
[]
>>> glob.glob('empty/**', recursive=True)
['empty/']
```

This happens in bash 5.0 too:

```
$ ls empty/*
ls: cannot access 'empty/*': No such file or directory
```

(note that the shell is matching nothing, so the `empty/*` argument is passed unchanged to `ls`)

Reply:
Aah, yes. So it's `fnmatch.translate` that's probably missing this expectation. I think I agree the bash/glob behavior is preferable, though it does complicate the implementation.